pkg/apis/garden/v1beta1/defaults.go

```diff
@@ -191,6 +191,15 @@ func setDefaults_ShootKubeControllerManager(kubernetes *Kubernetes) {
 	if hpa.UpscaleDelay == nil {
 		hpa.UpscaleDelay = &GardenerDuration{Duration: DefaultHPAUpscaleDelay}
 	}
+	if hpa.DownscaleStabilization == nil {
```

Comment:
The defaulting cannot happen this way, it must be happen depending on the Kubernetes version.
For all clusters we should default `SyncPeriod` and `Tolerance`.
For clusters < 1.12 we should default `DownscaleDelay` and `UpscaleDelay`.
For clusters >= 1.12 we should default `DownscaleStabilization`, `InitialReadinessDelay`, and `CPUInitializationPeriod`

Reply:
Perhaps something like

```
	if hpa.SyncPeriod == nil {
		hpa.SyncPeriod = &GardenerDuration{Duration: DefaultHPASyncPeriod}
	}
	if hpa.Tolerance == nil {
		defaultTolerance := DefaultHPATolerance
		hpa.Tolerance = &defaultTolerance
	}

	k8sVersionLessThan112, _ := utils.CompareVersions(kubernetes.Version, "<", "1.12")
	if k8sVersionLessThan112 {
		if hpa.DownscaleDelay == nil {
			hpa.DownscaleDelay = &GardenerDuration{Duration: DefaultHPADownscaleDelay}
		}
		if hpa.UpscaleDelay == nil {
			hpa.UpscaleDelay = &GardenerDuration{Duration: DefaultHPAUpscaleDelay}
		}
	} else {
		if hpa.DownscaleStabilization == nil {
			hpa.DownscaleStabilization = &GardenerDuration{Duration: DefaultDownscaleStabilization}
		}
		if hpa.InitialReadinessDelay == nil {
			hpa.InitialReadinessDelay = &GardenerDuration{Duration: DefaultInitialReadinessDelay}
		}
		if hpa.CPUInitializationPeriod == nil {
			hpa.CPUInitializationPeriod = &GardenerDuration{DefaultCPUInitializationPeriod}
		}
	}
```

Reply:
@amshuman-kr we agreed to not to the defaulting at all anymore because it will lead to too many problems with Kubernetes version constraints, and there is not a good way to react if the `CompareVersion` check fails. We argued that we also don't default other fields like that for which we have defaults in the Gardener, e.g. used `featureGates` or `runtimeConfig`. Are u OK with that?

Reply:
I am fine with that as long as we also adjust the Helm Chart templates to check and apply the command-line flags only if the values are specified in the spec.

Reply:
Yes, that's anyway happening implictlly already. Cool.

Reply:
> Yes, that's anyway happening implictlly already.

But the defaults are coming from the Helm Chart `values.yaml` which are different from `kube-controller-manager` defaults for some fields. This could be confusing for the users. WDYT?